staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -2105,9 +2105,9 @@ type PersistentVolumeClaimTemplate struct {
 
 // ContainerPort represents a network port in a single container.
 type ContainerPort struct {
-	// If specified, this must be an IANA_SVC_NAME and unique within the pod. Each
-	// named port in a pod must have a unique name. Name for the port that can be
-	// referred to by services.
+	// If specified, this must be an IANA_SVC_NAME. Each named port in a pod
+	// must have a unique name, otherwise, if there are duplicate names, only the
+	// first port name referred to by services will be chosen.
```

Comment:
@thockin @aojea is that consistently the behavior for endpoints, endpointslices, services, etc, or do we need to make the warning more ambiguous about potentially undefined behavior?

Reply:
ok. let me do the due diligence 

Services specify a TargetPort to map the Service Port to the Container Ports

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/apis/core/types.go#L4466-L4472

Endpoints and EndpointSlices reconcile this information in a single object

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/apis/core/types.go#L4599-L4602

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/apis/discovery/types.go#L148-L156

The endpoint and endpointslice controller use this `findPort` function for doing the mapping between service.TargetPort and container port

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/api/v1/pod/util.go#L28-L49

the endpointslice controller duplicated it to move the logic to staging

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/staging/src/k8s.io/endpointslice/utils.go#L378-L399

The controllers map the TargetPort to the containerPort , since the function is linear search, when a name is specified only the first pod is selected

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/controller/endpoint/endpoints_controller.go#L441-L449


https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/staging/src/k8s.io/endpointslice/utils.go#L84-L101


However, I could find that probes are also using the container port name

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/probe/util.go#L27C20-L56

so HTTPGetAction and TCPSocketActions probes are also impacted

https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/apis/core/types.go#L2201-L2207


https://github.com/kubernetes/kubernetes/blob/0cabb55f7c05eb269ae6f7292271833f095657ec/pkg/apis/core/types.go#L2231-L2238

hence, we must do it generic, I suggest something as

`first port name referred to by Services, Probes, ... will be chosen`

Reply:
something like "services and probes that select ports by name will use the first port with that name" sounds good to me

Reply:
> services and probes that select ports by name will use the first port with that name

I think that's better.
If we don't use an affirmative tone, it will cause confusion for users, who will need to understand the details of the design in more detail before they can use it.